### DISCUSSION TITLE
Narrowing of type Unknown

### DISCUSSION THREAD

**Discussion by haarisr** [microsoft/pyright#8094] [Q&A]
```python3
def foo(x, *args):
    return args


y: int
y = foo(3) # type of y is partially unknown, type of y in unknown|int
```

You could have  a type ignore comment but that does not change the type of y. 
This can also be avoided with `cast` but that has the overhead of function call (thought negligile) Sprinkling casts everywhere is not something that I would prefer.

We could just have a force cast statement or an ignore statment that force casts it into the desired type.

**Comment by erictraut** [comment]
I'm not sure what question you're asking here. Are you asking whether the behavior here is intentional or a bug? If so, the answer is that it's intentional.

It appears you have enabled the `reportUnknownVariableType` check, which is off by default. This check determines whether a variable is assigned a value with an `Unknown` type. In your example, the expression `foo(3)` has an unknown type, so it is appropriate for `reportUnknownVariableType` to generate an error.

**Reply by haarisr** [reply]
But what should the type be after `pyright: ignore`. I would assume `y` to be type `int` and not `int|unknown`. If this is intentional, is there any way to cast y to an int without the explict `cast` function.

Maybe `pyright` could add a `pyright: override` to force cast without the using the cast function

**Reply by erictraut** [reply]
> what should the type be after `pyright: ignore`?

I don't see a `pyright: ignore` in your example. In any case, `# pyright: ignore` has no impact on type evaluations. It simply suppresses diagnostics.

> is there any way to cast y to an int

You can apply an explicit cast to the expression `foo(3)` before its value is assigned to `y`.

```python
y = cast(int, foo(3))
```

I'll note that in your example, `foo(3)` does not return a value of type `int`. It returns `tuple[Unknown, ...]`, so this `cast` would serve to mask a bug.

Strict-mode checks like `reportUnknownVariableType` are most useful when your code is fully typed. In your code example, you have an unannotated function `foo`. I realize this is just a toy sample, but in your real code, you should add type annotations to the function that `foo` is meant to represent. If this function is in a library, then you may need to create a stub file.
